ztp: add Argo CD resource hook container image

vitus133 opened the pull request:
Adds container image, examples and instructions on building site installation pipeline with Argo CD

vitus133 commented:
/cc @imiller0 @nishant-parekh @yrobla

yuvalk commented:
/close

yuvalk commented:
/reopen
sorry, I meant to close a different pr

yuvalk commented on ztp/argocd/Containerfile:
```diff
@@ -0,0 +1,32 @@
+FROM quay.io/openshift/origin-cli:latest
```
If you want golang, I think it's better to use the openshift golang container base.

but in reality, I think you need golang as the build container. not in the target.
target should be as clean as possible so it's both small sized and reduce the security attack surface.

vitus133 replied:
I used origin-cli since I needed the oc /kubectl cli inside the container. @nishant-parekh is working on "golang AS BUILDER" thing, a PR will follow when ready

yuvalk commented on ztp/argocd/README.md:
```diff
@@ -0,0 +1,88 @@
+# README #
+
+This directory provides RAN site ArgoCD gitops pipeline container image definitions
```
you call it argocd,
but in reality there's nothing argocd specific about it IIUC. it's a container with the policygenerator
right?

vitus133 replied:
The concept of sync, pre-sync and post-sync resource hooks and examples provided are currently argocd specific

yuvalk commented on ztp/argocd/resource-hooks/pre-sync.yaml:
```diff
@@ -0,0 +1,21 @@
+    spec:
+      containers:
+        - name: ztp-site-generator
+          image: quay.io/redhat_emp1/ztp-site-generator:latest 
```
what's `redhat_emp1`?

vitus133 replied:
The Red Hat employee organization in Quay that we all can join to push / pool the same images, including private repos. See the Makefile for details

serngawy commented on ztp/argocd/deployment/argocd-controller-role-binding.yaml:
```diff
@@ -0,0 +1,12 @@
+apiVersion: rbac.authorization.k8s.io/v1
```
we already have it, under https://github.com/serngawy/cnf-features-deploy/blob/master/ztp/gitops-subscriptions/argocd/gitops-cluster-rolebinding.yaml

vitus133 replied:
fixed

serngawy commented on ztp/argocd/deployment/resource-hook-role-binding.yaml:
```diff
@@ -0,0 +1,13 @@
+apiVersion: rbac.authorization.k8s.io/v1
```
Please move all the files under ztp/argocd/  to ztp/gitops-subscriptions/argocd

vitus133 replied:
fixed

serngawy commented on ztp/argocd/deployment/resource-hook-role-binding.yaml:
```diff
@@ -0,0 +1,13 @@
+kind: ClusterRoleBinding
+metadata:
+  name: converter
+  namespace: ran-sites 
```
use NS clusters-sub as the clusters-app https://github.com/serngawy/cnf-features-deploy/blob/master/ztp/gitops-subscriptions/argocd/clusters-app.yaml

vitus133 replied:
fixed

serngawy commented on ztp/argocd/deployment/resource-hook-cluster-role.yaml:
```diff
@@ -0,0 +1,147 @@
+kind: ClusterRole
```
please move under ztp/gitops-subscriptions/argocd

vitus133 replied:
fixed

serngawy commented on ztp/argocd/resource-hooks/kustomization.yaml:
```diff
@@ -0,0 +1,4 @@
+resources:
```
please move this under ztp/gitops-subscriptions/argocd and better to sub directory as ztp/gitops-subscriptions/argocd/example and add test-sno.yaml there

vitus133 replied:
fixed

serngawy commented on ztp/argocd/src/common.sh:
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash
```
the ztp/argocd/src is not locked to Argocd any operator has add plugin feature can execute the container or the binaries. So its better to have the directory name something like ztp/ztp-resource-generator/src or ztp/ztp-cr-generator/src.

vitus133 replied:
fixed

vitus133 commented on ztp/argocd/Containerfile:
```diff
@@ -0,0 +1,32 @@
+    yum install -y --setopt=tsflags=nodocs $INSTALL_PKGS && \
+    rpm -V $INSTALL_PKGS && \
+    yum clean all -y && \
+    curl -sSLo - https://github.com/kubernetes-sigs/kustomize/releases/download/kustomize/v4.1.3/kustomize_v4.1.3_linux_amd64.tar.gz | \
```
Kustomize version embedded in the oc does not support what we need. 
We are already discussing ways to drop the use of kustomize inside of this container, so I think this will be the way forward for support.
@nishant-parekh

vitus133 commented on ztp/gitops-subscriptions/argocd/deployment/app-project.yaml:
```diff
@@ -0,0 +1,17 @@
+metadata:
+  name: default
+  namespace: openshift-gitops
+spec:
+  clusterResourceWhitelist:
+  - group: '*'
+    kind: '*'
+  destinations:
+  - namespace: '*'
+    server: '*'
+  namespaceResourceWhitelist:
+  - group: '*'
+    kind: '*'
+  sourceRepos:
+  - '*'
```
I changed to the permissions we actually used for the demo

serngawy replied:
Yes, we may need to have 2 different AppProjects one for clusetrs-app and other for common, groups and sites Apps  but let's merge your change first and then we can update.

imiller0 replied:
We should open a story to revisit these permissions and reduce them down to what is actually required.

vitus133 replied:
https://issues.redhat.com/browse/CNF-2727

vitus133 commented:
@serngawy I think I fixed all the items you pointed, could you please revisit? Also, I changed the ArgoCD project definition to the one we actually use in the demo. Also, do you think it worth defining a project other than the existing `default`, or maybe separating projects for day 1 and day2?

serngawy commented on ztp/gitops-subscriptions/argocd/deployment/kustomization.yaml:
```diff
@@ -8,3 +8,6 @@ resources:
   - common-app.yaml
   - groups-app.yaml
   - sites-app.yaml
+  - resource-hook-cluster-role.yaml
```
missing the clusters-app.yaml

vitus133 replied:
Nope, it's there - line 7

serngawy commented on ztp/gitops-subscriptions/argocd/deployment/resource-hook-role-binding.yaml:
```diff
@@ -0,0 +1,13 @@
+kind: ClusterRoleBinding
+metadata:
+  name: converter
+  namespace: sites-sub 
```
namespace must be clusters-sub. sites-sub used by sites-app which is for site specific policies.

vitus133 replied:
fixed

serngawy commented on ztp/gitops-subscriptions/argocd/deployment/sites-app.yaml:
```diff
@@ -16,7 +16,7 @@ spec:
     directory:
       recurse: true
-    path: ztp/ztp-policy-generator/out/sites
+    path: ztp/gitops-subscriptions/argocd/resource-hook-example
```
You mixing clusters-app with sites-app. Clusters-app for siteconfigs AND Sites-app for sites specific policies

vitus133 replied:
fixed

serngawy commented on ztp/gitops-subscriptions/argocd/deployment/resource-hook-role-binding.yaml:
```diff
@@ -0,0 +1,13 @@
+subjects:
+- kind: ServiceAccount
+  name: default
+  namespace: sites-sub
```
same

vitus133 replied:
fixed

serngawy commented on ztp/gitops-subscriptions/argocd/resource-hook-example/post-sync.yaml:
```diff
@@ -0,0 +1,21 @@
+kind: Job
+metadata:
+  name: siteconfig-post
+  namespace: sites-sub
```
clusters-sub

vitus133 replied:
Fixed

vitus133 commented:
@serngawy thanks for your remarks, fixed now

imiller0 commented on ztp/gitops-subscriptions/argocd/deployment/resource-hook-cluster-role.yaml:
```diff
@@ -0,0 +1,147 @@
+  - update
+
+
+- apiGroups:
```
I believe you can collapse these apiGroups into one:
- apiGroups:
  resources:
  - configmaps
  - namespaces
  - secrets
  ..

vitus133 replied:
Yes, but I noticed in different operator examples that this is not a common practice. Probably when it isn't collapsed it's less error prone, less of a chance to affect something unwanted when changing stuff

imiller0 commented on ztp/gitops-subscriptions/argocd/deployment/resource-hook-cluster-role.yaml:
```diff
@@ -0,0 +1,147 @@
+kind: ClusterRole
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: converter
```
Minor request. Maybe update the name to be more reflective of the function, eg site-converter

vitus133 replied:
done

imiller0 commented on ztp/resource-generator/Containerfile:
```diff
@@ -0,0 +1,32 @@
+WORKDIR /usr/src/hook
+# For Python client
+RUN python -m venv venv && \
+    source venv/bin/activate && \
```
Since this is in a container is a virtualenv needed or can kubernetes and jinja2 be installed directly in the container OS?

vitus133 replied:
https://issues.redhat.com/browse/CNF-2728

imiller0 commented on ztp/resource-generator/README.md:
```diff
@@ -0,0 +1,88 @@
+## What is this repository for? ##
+### Quick summary ###
+
+The idea is to accept RAN site manifests in a format defined by "siteconfig" CRD, convert them to custom resources accepted by the assisted service and ACM, and apply them to the cluster running Assisted installer operator and ACM hub.
```
Replace "siteconfig" with SiteConfig
Assisted installer is the Assisted Service Operator

imiller0 commented on ztp/resource-generator/README.md:
```diff
@@ -0,0 +1,88 @@
+2. Red Hat GitOps operator installed
+3. Red Hat Advanced cluster Management for Kubernetes installed (version 2.3 or above)
+4. If you intend to build images yourself:
+   1. To push images you build, you need [quay.io](quay.io) credentials and belonging to [Red Hat employees organization in quay](https://quay.io/organization/redhat_emp1). But you can always fork this repo and push to your private account or another registry if desired  (Change the `REG ?= quay.io/redhat_emp1/ztp-site-generator` line in the [Makefile](Makefile) to the container registry of your  ...
```
Also note that they need to update the pre and post hooks to point to their image.

imiller0 commented on ztp/resource-generator/README.md:
```diff
@@ -0,0 +1,88 @@
+### Project organization ###
+#### [resource-hooks](../gitops-subscriptions/argocd/resource-hook-example) ####
+The [resource-hooks](../gitops-subscriptions/argocd/resource-hook-example) directory contains an example content of a site configuration repository that we want to define in the ArgoCD application to synchronize with our hub cluster.
+- [test-sno.yaml](../gitops-subscriptions/argocd/resource-hook-example/test-sno.yaml) - site definitions
```
Make a note that the customer will replicate this file for each new site.

imiller0 commented on ztp/resource-generator/README.md:
````diff
@@ -0,0 +1,88 @@
+```
+
+### Example setup ###
+1. In a terminal, clone this repo and change the working derectory to this one
````
Change to which working directory?

vitus133 commented:
@imiller0  thanks for your remarks, fixed now

serngawy commented on ztp/resource-generator/README.md:
```diff
@@ -0,0 +1,88 @@
+# README #
+
+This directory provides RAN site ArgoCD gitops pipeline container image definitions
```
I think its better to explain the resource-generator without tying it to ArgoCD. Its main purpose to use policyGen in container image to generate CRs.

vitus133 replied:
The concept of pre-sync and post-sync resource hooks is unique to the ArgoCD pipeline, so I think it's more than just packing the policy generator into a container. Maybe the resource-generator folder should be renamed?

serngawy commented on ztp/resource-generator/README.md:
```diff
@@ -0,0 +1,88 @@
+## What is this repository for? ##
+### Quick summary ###
+
+The idea is to accept RAN site manifests in a format defined by "SiteConfig" CRD, convert them to custom resources accepted by the assisted service and ACM, and apply them to the cluster running Assisted Service Operator and ACM hub.
```
The plain is to use the same container to generate the acm-policies using PolicyGenTemplate CRD. 

Are you planing to update this readme later ? sounds like it follows the old directory structure where everything under the same argocd dir. 

I feel Its better to make this readme only for how to build and execute the image ? what do u think ?

vitus133 replied:
I hear you. And yet I think this or similar content should be defined somewhere. Maybe when we do the changes we discussed for policy generator, this directory will be bound to to a massive overhaul?

lack commented on ztp/gitops-subscriptions/argocd/resource-hook-example/test-sno.yaml:
```diff
@@ -0,0 +1,67 @@
+  pullSecretRef:
+    name: "assisted-deployment-pull-secret"
+  clusterImageSetNameRef: "openshift-4.8"
+  sshPublicKey: "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDB3dwhI5X0ZxGBb9VK7wclcPHLc8n7WAyKjTNInFjYNP9J+Zoc/ii+l3YbGUTuqilDwZN5rVIwBux2nUyVXDfaM5kPd9kACmxWtfEWTyVRootbrNWwRfKuC2h6cOd1IlcRBM1q6IzJ4d7+JVoltAxsabqLoCbK3svxaZoKAaK7jdGG030yvJzZaNM4PiTy39VQXXkCiMDmicxEBwZx1UsA8yWQsiOQ5brod9KQRXWAAST779gbvtgXR2L+MnVNROEHf1nEjZJwjwaHxoDQYHYKERxKRHlWFtmy5dNT6BbvOpJ2e5osDFPMEd41d2mUJTfxXiC1nvyjk9Irf8YJYnqJgB ...
```
It's messy, but public keys are safe to have in source control.

vitus133 commented:
Thanks folks for the comments!
I'm closing this CR in favor of the one that includes Day-2 operations. Coming soon.

vitus133 commented:
Decided with the team to reopen it

yrobla commented on ztp/resource-generator/README.md:
```diff
@@ -0,0 +1,88 @@
+4. If you intend to build images yourself:
+   1. To push images you build, you need [quay.io](quay.io) credentials and belonging to [Red Hat employees organization in quay](https://quay.io/organization/redhat_emp1). But you can always fork this repo and push to your private account or another registry if desired  (Change the `REG ?= quay.io/redhat_emp1/ztp-site-generator` line in the [Makefile](Makefile) to the container registry of your  ...
+   2. podman installed
+   3. make installed
```
maybe mention that you need a git server, reachable from the hub?

vitus133 replied:
Done

yrobla reviewed (changes requested):
small typo

vitus133 commented on ztp/resource-generator/README.md:
```diff
@@ -0,0 +1,89 @@
+   1. To push images you build, you need [quay.io](quay.io) credentials and belonging to [Red Hat employees organization in quay](https://quay.io/organization/redhat_emp1). But you can always fork this repo and push to your private account or another registry if desired  (Change the `REG ?= quay.io/redhat_emp1/ztp-site-generator` line in the [Makefile](Makefile) to the container registry of your  ...
+   2. podman installed
+   3. make installed
+   4. For disconnected environments please make sure your repository of SiteConfig resources and atp-site-generator container image are accessible from the hub cluster.
```
Done, sorry

vitus133 commented:
/test e2e-gcp-ran-profile

yrobla commented:
/lgtm

serngawy commented:
/retest

nishant-parekh commented:
/retest

vitus133 commented:
/test e2e-gcp-ovn

nishant-parekh commented:
/retest

vitus133 commented:
/hold

vitus133 commented:
/unhold

serngawy commented:
/lgtm

vitus133 commented:
/test e2e-gcp-ran-profile
